assertsprocessor/service_queue.go

```diff
@@ -0,0 +1,64 @@
+package assertsprocessor
+
+import (
+	"math"
+	"sync"
+)
+
+type serviceQueues struct {
+	config        *Config
+	requestStates *sync.Map
+	requestCount  int
+	rwMutex       *sync.RWMutex
+}
+
+func NewServiceQueues(config *Config) *serviceQueues {
+	return &serviceQueues{
+		config:        config,
+		requestStates: &sync.Map{},
+		rwMutex:       &sync.RWMutex{},
+	}
+}
+
+func (sq *serviceQueues) getRequestState(request string) *traceSampler {
+	entry, found := sq.requestStates.Load(request)
+	if found {
+		return entry.(*traceSampler)
+	}
+
+	// We create an entry for a request only when there is room
+	var result *traceSampler
+
+	// Check if there is room
+	if sq.hasRoom() {
```

Comment:
The double-checked locking approach here can guarantee only one entry is created per request, but it cannot strictly enforce the `limitPerService`.

Suppose the `currentSize` is exactly `limitPerService - 1`, two threads processing two different requests can finish `hasRoom` calls with the return value being true. Even though there is a lock at the beginning of the following block, it doesn't prevent the `requestCount` from being bumped to `limitPerService + 1`. The two threads will just take turns executing the following block.

Reply:
Good catch. I had the check before but removed it. Will bring it back